pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/cleanup/StaleInstancesCleanupTask.java

```diff
@@ -82,7 +95,47 @@ protected void runTask(Properties periodicTaskProperties) {
             _controllerMetrics.addValueToGlobalGauge(ControllerGauge.DROPPED_MINION_INSTANCES, 1);
           }
         }
+        continue;
+      }
+
+      // Drop the broker instance if it has been offline for more than a period of this task.
+      if (InstanceTypeUtils.isBroker(offlineInstance) && !brokerInstancesInUse.contains(offlineInstance)) {
+        if (_pinotHelixResourceManager.isInstanceOfflineFor(offlineInstance,
+            _staleInstancesCleanupTaskMinOfflineTimeBeforeDeletionInMilliseconds)) {
+          LOGGER.info("Dropping broker instance: {}", offlineInstance);
+          PinotResourceManagerResponse response = _pinotHelixResourceManager.dropInstance(offlineInstance);
+          if (response.isSuccessful()) {
+            _controllerMetrics.addValueToGlobalGauge(ControllerGauge.DROPPED_BROKER_INSTANCES, 1);
+          }
+        }
+        continue;
+      }
+
+      // Drop the server instance if it has been offline for more than a period of this task.
+      if (InstanceTypeUtils.isServer(offlineInstance) && !serverInstancesInUse.contains(offlineInstance)) {
+        if (_pinotHelixResourceManager.isInstanceOfflineFor(offlineInstance,
+            _staleInstancesCleanupTaskMinOfflineTimeBeforeDeletionInMilliseconds)) {
+          LOGGER.info("Dropping server instance: {}", offlineInstance);
+          PinotResourceManagerResponse response = _pinotHelixResourceManager.dropInstance(offlineInstance);
+          if (response.isSuccessful()) {
+            _controllerMetrics.addValueToGlobalGauge(ControllerGauge.DROPPED_SERVER_INSTANCES, 1);
+          }
+        }
       }
     }
   }
+
+  private Set<String> getBrokerInstancesInUse() {
```

Comment:
No need to read all tables. Only need to read the `brokerResource` IS.

Reply:
updated to use brokerResource